### ISSUE TITLE
Restructuring dynamic shadow masking

### ISSUE THREAD

**Issue by jmp1985** [dials/dials#705]
This ticket follows a discussion about mask handling with @graeme-winter 

Currently, masking is a bit messy in dials. We have the ability to associate a static 2D mask during the import step; however, a mask can also be set (to possibly override?) during spot finding and integration. We also have the ability for masking to be done in the format class through the get_mask method. This method may perform dynamic shadow masking or it may just return the trusted range mask based on the image pixel values. 

The problem with the get_mask method is that because the masking is implemented in python, it can not easily be called from a separate thread. This means that during the threaded integration, the get_mask method takes up a large amount of the processing time. 

The proposed solution to change where the masking is performed. The dynamic mask only needs to be computed once; however, we compute it on the fly every time we read the image; this takes up unnecessary CPU cycles. At Diamond beamlines (particularly for EIGER data which is already stored in HDF5 files) it may be possible to have the dynamic mask generated during data collection. For other data we could generate the mask during import. 

A format class has a method which tells us whether dynamic masking is available so dials.import can check whether the format class requires a dynamic mask. Then if a parameter (e.g. generate_mask=True) is set, the mask can be generated and stored in an external HDF5 file and linked to the experiments.json file. If generate_mask=False then a warning can be issued to the user telling them that they should generate a mask but that it may take some time. This would mean that user's using the dials command lines might have to do an additional step but xia2 would do it automatically.

`dials.import *.cbf generate_mask=True`

Then during spot finding and integration, the mask associated with the experiments list file would be used. If the mask is a 2D array then it is treated as a static mask. If it is a 3D array then it is treated as a dynamic mask.

If the mask array has 0 for a masked out pixel and 1 for a valid pixel, then the mask would be applied to the image pixels as follows: 

`image[j,i] = mask[j,i] * image[j,i] + mask_code * (!mask[j,i])`

1. image[j,i] = C; mask[j,i] = 0 ---> image[j,i] = 0*C+mask_code*1 = mask_code
2. image[j,i] = C; mask[j,i] = 1 ---> image[j,i] = 1*C+mask_code*0 = C

where mask_code might be some negative number.

The mask file would be an HDF5 file with the following entry (or something similar)

`/entry/shadow`

We could actually have versioned masks,  for example a file might contain the following datasets with a hardlink to the desired mask

```
/entry/shadow -> /entry/shadow-0003
/entry/shadow-0001
/entry/shadow-0002
/entry/shadow-0003
```

This would mean that the image viewer mask tool could update the mask file with the new mask it creates. For example, lets assume a mask has been generated during import. Now a user makes a new mask in the image viewer and saves it. This creates a new version in the same file. The user then changes their mind and opens up the image viewer and can now go back to the original mask to make changes to that (or work off their initial modified version). 

Since the mask arrays would be sparse 0,1 values arrays they would compress very well so initial tests by Graeme indicate they wouldn't take up a lot of space. 

One thing to consider is that, often a mask is created in the image viewer just for spot finding but not necessarily for use in integration. In this case, some way of overriding the mask may still be necessary. 

In conclusion, doing masking up front has major advantages for the data processing and won't add any overall overhead. Although it might make import slower.

**Comment by rjgildea** [comment]
Resolved by https://github.com/cctbx/dxtbx/pull/59
